crates/oxc_linter/src/rules/promise/no_callback_in_promise.rs

```diff
@@ -0,0 +1,192 @@
+    AstKind,
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{CompactStr, GetSpan, Span};
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+fn no_callback_in_promise_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Avoid calling back inside of a promise")
+        .with_help("Use `then` and `catch` directly, avoid callbacks")
+        .with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoCallbackInPromise(Box<NoCallbackInPromiseConfig>);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoCallbackInPromiseConfig {
+    callbacks: Vec<CompactStr>,
+}
+
+impl std::ops::Deref for NoCallbackInPromise {
+    type Target = NoCallbackInPromiseConfig;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallows calling a callback function (`cb()`) inside a `Promise.prototype.then()`
+    /// or `Promise.prototype.catch()`.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Directly invoking a callback inside a `then()` or `catch()` method can lead to
+    /// unexpected behavior, such as the callback being called multiple times. Additionally,
+    /// mixing the callback and promise paradigms in this way can make the code confusing
+    /// and harder to maintain.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// function callback(err, data) {
+    ///     console.log('Callback got called with:', err, data)
+    ///     throw new Error('My error')
+    ///   }
+    ///
+    /// Promise.resolve()
+    ///   .then(() => callback(null, 'data'))
+    ///   .catch((err) => callback(err.message, null))
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// Promise.resolve()
+    ///   .then((data) => { console.log(data) })
+    ///   .catch((err) => { console.error(err) })
+    /// ```
+    NoCallbackInPromise,
+    correctness,
+);
+
+impl Rule for NoCallbackInPromise {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let default_callbacks: Vec<CompactStr> =
+            ["done", "cb", "callback", "next"].map(Into::into).to_vec();
+
+        let exceptions: Vec<String> = value
+            .get(0)
+            .and_then(|v| v.get("exceptions"))
+            .and_then(serde_json::Value::as_array)
+            .map(|v| {
+                v.iter().filter_map(serde_json::Value::as_str).map(ToString::to_string).collect()
+            })
+            .unwrap_or_default();
+
+        Self(Box::new(NoCallbackInPromiseConfig {
+            callbacks: default_callbacks
+                .into_iter()
+                .filter(|item| !exceptions.contains(&item.to_string()))
+                .collect(),
+        }))
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if !self.is_callback(node) {
```

Comment:
You can move the code in this function into this rule, then reference the call_expr without needing to unwrap later